domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,305 @@ class TopicListController @Inject constructor(
+      )
+      .addAllUpTo(
+        populateOlderPlayedStories(topicProgressList),
+        PromotedStoryList.Builder::addAllOlderPlayedStory,
+        limit = 3
+      )
+      .addAllUpTo(
+        computeSuggestedStories(topicProgressList),
+        PromotedStoryList.Builder::addAllSuggestedStory,
+        limit = 3
+      ).build()
+  }
+
+  private fun PromotedStoryList.getTotalPromotedStoryCount(): Int {
+    return recentlyPlayedStoryList.size + olderPlayedStoryList.size + suggestedStoryList.size
+  }
+
+  private fun PromotedStoryList.Builder.getTotalPromotedStoryCount(): Int {
+    return recentlyPlayedStoryList.size + olderPlayedStoryList.size + suggestedStoryList.size
+  }
+
+  private fun PromotedStoryList.Builder.addAllUpTo(
+    iterable: Iterable<PromotedStory>,
+    addAll: PromotedStoryList.Builder.(Iterable<PromotedStory>) -> PromotedStoryList.Builder,
+    limit: Int
+  ): PromotedStoryList.Builder {
+    return this.addAll(iterable.take(limit - this.getTotalPromotedStoryCount()))
+  }
+
+  private fun populateRecentlyPlayedStories(
+    topicProgressList: List<TopicProgress>
+  ): List<PromotedStory> {
+    computePromotedStories(topicProgressList).let {
+      Pair(it.first, it.second)
+      if (it.second < ONE_WEEK_IN_DAYS)
+        return it.first
+    }
+    return emptyList()
+  }
+
+  private fun populateOlderPlayedStories(
     topicProgressList: List<TopicProgress>
-  ): OngoingStoryList {
-    val ongoingStoryListBuilder = OngoingStoryList.newBuilder()
-    topicProgressList.forEach { topicProgress ->
+  ): List<PromotedStory> {
+    computePromotedStories(topicProgressList).let {
+      Pair(it.first, it.second)
+      if (it.second > ONE_WEEK_IN_DAYS)
+        return it.first
+    }
+    return emptyList()
+  }
+
+  private fun computePromotedStories(
+    topicProgressList: List<TopicProgress>
+  ): Pair<List<PromotedStory>, Long> {
+    var numberOfDaysPassed = 0L
+    val recentlyPlayedPromotedStoryList = mutableListOf<PromotedStory>()
+    val sortedTopicProgressList =
+      topicProgressList.sortedByDescending { it.lastPlayedTimestamp }
+
+    sortedTopicProgressList.forEach { topicProgress ->
       val topic = topicController.retrieveTopic(topicProgress.topicId)
+
       topicProgress.storyProgressMap.values.forEach { storyProgress ->
         val storyId = storyProgress.storyId
         val story = topicController.retrieveStory(topic.topicId, storyId)
 
-        val completedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
-        val lastCompletedChapterProgress: ChapterProgress? =
+        val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
+        val mostRecentCompletedChapterProgress: ChapterProgress? =
```

Comment:
Is this most recent or least recent? How is the returned chapter progress list sorted?

Reply:
It is `latestCompletedChapterProgress` .